app/streaming/video/ffmpeg-renderers/eglvid.cpp

```diff
@@ -0,0 +1,480 @@
+                 "EGLRenderer: cannot create shader program");
+        goto prog_fail_create;
+    }
+
+    glAttachShader(m_shader_program, vertex_shader);
+    glAttachShader(m_shader_program, fragment_shader);
+    glLinkProgram(m_shader_program);
+    int status;
+    glGetProgramiv(m_shader_program, GL_LINK_STATUS, &status);
+    if (status) {
+        ret = true;
+    } else {
+        char shader_log[512];
+        glGetProgramInfoLog(m_shader_program, sizeof (shader_log), nullptr, shader_log);
+        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION,
+                 "EGLRenderer: cannot link shader program: %s",
+                 shader_log);
+        glDeleteProgram(m_shader_program);
+    }
+
+prog_fail_create:
+    glDeleteShader(fragment_shader);
+frag_error:
+    glDeleteShader(vertex_shader);
+    return ret;
+}
+
+bool EGLRenderer::initialize(PDECODER_PARAMETERS params)
+{
+    m_window = params->window;
+
+    if (params->videoFormat == VIDEO_FORMAT_H265_MAIN10) {
+        // SDL doesn't support rendering YUV 10-bit textures yet
+        // TODO: FIXME
+        return false;
+    }
+
+    SDL_SysWMinfo info;
+    SDL_VERSION(&info.version);
+    if (!SDL_GetWindowWMInfo(params->window, &info)) {
+        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION,
+             "!SDL_GetWindowWMInfo() failed: %s",
+             SDL_GetError());
+        return false;
+    }
+    switch (info.subsystem) {
+    case SDL_SYSWM_WAYLAND:
+        m_egl_display = eglGetPlatformDisplay(EGL_PLATFORM_WAYLAND_KHR,
+                                              info.info.wl.display, nullptr);
+        break;
+    case SDL_SYSWM_X11:
+        m_egl_display = eglGetPlatformDisplay(EGL_PLATFORM_X11_KHR,
+                                              info.info.x11.display, nullptr);
+        break;
+    default:
+        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION,
+                     "EGLRenderer: not compatible with SYSWM");
+        return false;
+    }
+
+    if (!m_egl_display) {
+        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Cannot get EGL display: ");
+        return false;
+    }
+
+    if (!(m_context = SDL_GL_CreateContext(params->window))) {
+        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Cannot create OpenGL context: %s",
+                     SDL_GetError());
+        return false;
+    }
+    if (SDL_GL_MakeCurrent(params->window, m_context)) {
+        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Cannot use created EGL context: %s",
+                     SDL_GetError());
+        return false;
+    }
+
+    const EGLExtensions egl_extensions(m_egl_display);
+    if (!egl_extensions.is_supported("EGL_KHR_image_base") &&
+        !egl_extensions.is_supported("EGL_KHR_image")) {
+        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "EGL: KHR_image unsupported...");
+        deinitialize();
+        return false;
+    }
+
+    if (!m_frontend->initializeEGL(m_egl_display, egl_extensions)) {
+        deinitialize();
```

Comment:
These `deinitialize()` calls are redundant. The destructor will take care of cleaning up after failure.

Reply:
`initialize` can only be called once (there will never be some retry on failure) ?